File: code/design-patterns/structural/facade/python/facade.py
class TV:
    def turn_on(self):
        print("TV is ON")

    def turnOff(self):
        print("TV is OFF")


class SoundSystem:
    def turn_on(self):
        print("Sound System is ON")

    def turnOff(self):
        print("Sound System is OFF")


class GamingConsole:
    def turn_on(self):
        print("Gaming Console is ON")

    def turnOff(self):
        print("Gaming Console is OFF")


class HomeEntertainmentFacade:

    def __init__(self):
        self.tv = TV()
        self.sound_system = SoundSystem()
        self.gaming_console = GamingConsole()

    def playGame(self):
        self.gaming_console.turn_on()
        self.tv.turn_on()
        self.sound_system.turn_on()

    def watchMovie(self):
        self.tv.turn_on()
        self.sound_system.turn_on()

File: code/design-patterns/structural/facade/go/facade.go
package facade

import "fmt"

// TV struct
type TV struct{}

func (t *TV) TurnOn() {
	fmt.Println("TV is ON")
}

func (t *TV) TurnOff() {
	fmt.Println("TV is OFF")
}

// SoundSystem struct
type SoundSystem struct{}

func (s *SoundSystem) TurnOn() {
	fmt.Println("Sound System is ON")
}

func (s *SoundSystem) TurnOff() {
	fmt.Println("Sound System is OFF")
}

// GamingConsole struct
type GamingConsole struct{}

func (g *GamingConsole) TurnOn() {
	fmt.Println("Gaming Console is ON")
}

func (g *GamingConsole) TurnOff() {
	fmt.Println("Gaming Console is OFF")
}

// HomeEntertainmentFacade struct
type HomeEntertainmentFacade struct {
	tv            *TV
	soundSystem   *SoundSystem
	gamingConsole *GamingConsole
}

func NewHomeEntertainmentFacade() *HomeEntertainmentFacade {
	return &HomeEntertainmentFacade{
		tv:            &TV{},
		soundSystem:   &SoundSystem{},
		gamingConsole: &GamingConsole{},
	}
}

func (h *HomeEntertainmentFacade) PlayGame() {
	h.gamingConsole.TurnOn()
	h.tv.TurnOn()
	h.soundSystem.TurnOn()
}

func (h *HomeEntertainmentFacade) WatchMovie() {
	h.tv.TurnOn()
	h.soundSystem.TurnOn()
}
